src/stan/math/prim/mat/meta/index_type.hpp

```diff
@@ -21,6 +21,16 @@ namespace stan {
       typedef typename Eigen::Matrix<T,R,C>::Index type;
     };
 
+    template <typename T>
+    struct index_type<Eigen::MatrixBase<T> > {
+      typedef typename index_type<T>::type type;
+    };
+
+    template <typename T>
+    struct index_type<Eigen::Map<T> > {
```

Comment:
Hmmm.... does that mean we need to have separate function signatures for `MatrixBase` and `Map`? In other words, do we to split every function that has `Eigen::Matrix` in the arguments into `MatrixBase` and `Map`?